test/examples/ucx_hello_world.h

```diff
@@ -35,6 +48,140 @@
+                    #_func, cudaGetErrorString(_result));  \
+        }                                                  \
+    } while(0)
+
+
+void *mem_type_malloc(size_t length)
+{
+    void *ptr;
+
+    switch (test_mem_type) {
+    case UCS_MEMORY_TYPE_HOST:
+        ptr = malloc(length);
+        break;
+#ifdef HAVE_CUDA
+    case UCS_MEMORY_TYPE_CUDA:
+        CUDA_FUNC(cudaMalloc(&ptr, length));
+        break;
+    case UCS_MEMORY_TYPE_CUDA_MANAGED:
+        CUDA_FUNC(cudaMallocManaged(&ptr, length, length));
+        break;
+#endif
+    default:
+        fprintf(stderr, "Unsupported memory type: %d", test_mem_type);
+        ptr = NULL;
+        break;
+    }
+
+    return ptr;
+}
+
+void mem_type_free(void *address)
+{
+    switch (test_mem_type) {
+    case UCS_MEMORY_TYPE_HOST:
+        free(address);
+        break;
+#ifdef HAVE_CUDA
+    case UCS_MEMORY_TYPE_CUDA:
+    case UCS_MEMORY_TYPE_CUDA_MANAGED:
+        CUDA_FUNC(cudaFree(address));
+        break;
+#endif
+    default:
+        fprintf(stderr, "Unsupported memory type: %d", test_mem_type);
+        break;
+    }
+}
+
+void *mem_type_memcpy(void *dst, const void *src, size_t count)
+{
+    switch (test_mem_type) {
+    case UCS_MEMORY_TYPE_HOST:
+    case UCS_MEMORY_TYPE_CUDA_MANAGED:
+        memcpy(dst, src, count);
+        break;
+#ifdef HAVE_CUDA
+    case UCS_MEMORY_TYPE_CUDA:
+        CUDA_FUNC(cudaMemcpy(dst, src, count, cudaMemcpyDefault));
+        break;
+#endif
+    default:
+        fprintf(stderr, "Unsupported memory type: %d", test_mem_type);
+        break;
+    }
+
+    return dst;
+}
+
+void *mem_type_memset(void *dst, int value, size_t count)
+{
+    switch (test_mem_type) {
+    case UCS_MEMORY_TYPE_HOST:
+    case UCS_MEMORY_TYPE_CUDA_MANAGED:
+        memset(dst, value, count);
+        break;
+#ifdef HAVE_CUDA
+    case UCS_MEMORY_TYPE_CUDA:
+        CUDA_FUNC(cudaMemset(dst, value, count));
+        break;
+#endif
+    default:
+        fprintf(stderr, "Unsupported memory type: %d", test_mem_type);
+        break;
+    }
+
+    return dst;
+}
+
+int check_mem_type_support(ucs_memory_type_t mem_type)
+{
+    switch (test_mem_type) {
+    case UCS_MEMORY_TYPE_HOST:
+        return 1;
+    case UCS_MEMORY_TYPE_CUDA:
+    case UCS_MEMORY_TYPE_CUDA_MANAGED:
+#ifdef HAVE_CUDA
+        return 1;
+#else
+        return 0;
+#endif
+    default:
+        fprintf(stderr, "Unsupported memory type: %d", test_mem_type);
+        break;
+    }
+
+    return 0;
+}
+
+int parse_mem_type(const char *opt_arg)
+{
+    if (!strcmp(opt_arg, "host")) {
+        test_mem_type = UCS_MEMORY_TYPE_HOST;
+    } else if (!strcmp(opt_arg, "cuda") &&
+               check_mem_type_support(UCS_MEMORY_TYPE_CUDA)) {
+        test_mem_type = UCS_MEMORY_TYPE_CUDA;
+    } else if (!strcmp(opt_arg, "cuda-managed") &&
+               check_mem_type_support(UCS_MEMORY_TYPE_CUDA_MANAGED)) {
+        test_mem_type = UCS_MEMORY_TYPE_CUDA_MANAGED;
```

Comment:
maybe return memory type instead of setting global var?
error could be MEM_TYPE_UNKNOWN :)

Reply:
done